### ISSUE TITLE
Add unit tests for IterativeClosestPointWithNormals

### ISSUE THREAD

**Issue by kunaltyagi** [PointCloudLibrary/pcl#3535]
> what needs to be checked is how this escaped surfacing during unit tests

Turns out we do not have any unit tests for `IterativeClosestPointWithNormals`.

_Originally posted by @taketwo in https://github.com/PointCloudLibrary/pcl/issues/3462#issuecomment-572758837_

**Comment by taketwo** [comment]
Also, the existing tests should be updated to test the classes instantiated with both `float` and `double` precision.

**Comment by kaipomauli** [comment]
Hi, 
I am interested in learning more about how the ICP algorithm works and as such would like to be assigned this issue. Could you outline where the test would be written, and what to the unit test would check for?
Thanks

**Comment by larshg** [comment]
> Hi, I am interested in learning more about how the ICP algorithm works and as such would like to be assigned this issue. Could you outline where the test would be written, and what to the unit test would check for? Thanks

Sounds great.

There seems to be something tested here:
https://github.com/PointCloudLibrary/pcl/blob/14a8bde65eb9a71711706db47c0269a40b91a8b4/test/registration/test_registration.cpp#L235-L244

But its not testing the actual alignment/registration. You could look at the other tests in here to get an idea.
Ie. have a source cloud, translate/rotate it a fixed amount and then use ICPWithNormals to find the alignment back again.
Verify that the points get close to the originally cloud.

**Comment by mvieth** [comment]
The easiest thing to do would be to copy `TEST (PCL, IterativeClosestPoint)` (would also be fine by me).
More interesting would be a test that shows the difference between ICP and ICPWithNormals. So perhaps ICPWithNormals gives more accurate results? Or there is a situation where ICP fails and gives a wrong alignment, but ICPWithNormals gives the correct alignment? To be honest, I am not sure how such a situation might look like.

**Comment by larshg** [comment]
> Or there is a situation where ICP fails and gives a wrong alignment, but ICPWithNormals gives the correct alignment?

I also thought about this, but it sounds more like a benchmark/comparison in precision/usecase - rather than just a unit test :)

**Comment by mvieth** [comment]
You are right, it should not be a direct comparison between the two. My thought was just: ICPWithNormals was probably added for a specific reason: to be more precise, or to give the correct result in specific situations, or whatever. Ideally, the test would reflect that, so that people who use ICPWithNormals because of that reason, are guaranteed to get the desired behaviour
But I am also happy with simply copying the other test

**Comment by kaipomauli** [comment]
Thank you for your suggestions I will dig and see what I find while testing the two.

**Comment by harshareddy1611** [comment]
Hi @kaipomauli, are you still working on this? If not, I'd be happy to pick it up.

@mvieth @larshg if it's open, my plan would be the agreed approach from the discussion above: add a real alignment test in `test/registration/test_registration.cpp` alongside the existing `TEST (PCL, IterativeClosestPointWithNormals)` — take a source cloud (with normals), apply a known translation/rotation, run `IterativeClosestPointWithNormals` to recover the transform, and assert the aligned points come back within tolerance of the original. I'll instantiate it with both `float` and `double` precision as taketwo suggested.

I'll keep it to the straightforward version for this PR. The "situation where normals give a better/correct result than plain ICP" idea is interesting but, as larshg noted, reads more like a benchmark — happy to look at that as a follow-up if useful.

Let me know and I'll get started.

**Comment by kaipomauli** [comment]
> Hi [@kaipomauli](https://github.com/kaipomauli), are you still working on this? If not, I'd be happy to pick it up.
> 
> [@mvieth](https://github.com/mvieth) [@larshg](https://github.com/larshg) if it's open, my plan would be the agreed approach from the discussion above: add a real alignment test in `test/registration/test_registration.cpp` alongside the existing `TEST (PCL, IterativeClosestPointWithNormals)` — take a source cloud (with normals), apply a known translation/rotation, run `IterativeClosestPointWithNormals` to recover the transform, and assert the aligned points come back within tolerance of the original. I'll instantiate it with both `float` and `double` precision as taketwo suggested.
> 
> I'll keep it to the straightforward version for this PR. The "situation where normals give a better/correct result than plain ICP" idea is interesting but, as larshg noted, reads more like a benchmark — happy to look at that as a follow-up if useful.
> 
> Let me know and I'll get started.

Hi no I am not! be my guest! real busy these days :(

**Comment by mvieth** [comment]
@harshareddy1611 Feel free to go ahead and open a pull request with your suggested test.
